FS: Reviews

jb162 opened the pull request:
### What this PR does (required):
- addresses #42 
- creates reviews container on profile details page, as well as review form to leave a review


### Screenshots / Videos (front-end only):
![image](https://user-images.githubusercontent.com/78042694/148841502-8d824af0-3e9a-4acf-b0c9-1a06c2e75c9d.png)


### Any information needed to test this feature (required):
- run server and app
- navigate to profile details page 
   - http://localhost:3000/profile-details/<sitter's user id>
- submit a review via the form on the right hand side
- after page reload, your review should appear under the sitter's profile details on the left-hand side
### Any issues with the current functionality (optional):
- currently does not display reviewer's name (need to add in fetchProfile call with clientId)
- so far, anyone can leave a review whether they have booked the sitter or not (can change this once everything is integrated)
- need to add in functionality on profile details page to change user's rating each time a new review is submitted for them

jb162 committed:
Initial Commit

need to rebase

jb162 committed:
Created getReviews helper

able to get reviews from back-end and set to state var. currently working on FE component (displaying reviews on profile details page)

jb162 committed:
Integration of Reviews

created helper to create new review. everything is fully functional, just need to put final styling touches

jb162 committed:
Reviews Container UI changes

updated reviews container styling

rajivtitus commented on client/src/helpers/APICalls/getReviews.ts:
```diff
@@ -0,0 +1,4 @@
+export const getReviews = async (sitterId: string) => {
+  const response = await fetch(`http://localhost:3000/reviews/all/${sitterId}`);
```
Nit: Can we please specify the options for the fetch client (please see handler above)

rajivtitus commented on client/src/pages/ProfileDetails/BookingForm/BookingForm.tsx:
```diff
@@ -1,69 +1,108 @@
   const formik = useFormik({
     initialValues: {
-      startDate: '',
+      startDate: 'placeholder',
```
Why don't we use the values from the default state above?

Ditto below!

jb162 replied:
I have the values separated since I don't want to pre-fill the form with any values returned by the fetchProfile call. Each of the initialValues maps to a MenuItem that displays the relevant placeholder for each input on the form.

rajivtitus commented on client/src/pages/ProfileDetails/BookingForm/useStyles.ts:
```diff
@@ -5,7 +5,8 @@ const useStyles = makeStyles((theme) => ({
     height: '55%',
-    justifySelf: 'flex-end',
+    minHeight: 550,
+    // justifySelf: 'flex-end',
```
Please uncomment or remove

rajivtitus commented on client/src/pages/ProfileDetails/ProfileDetails.tsx:
```diff
@@ -49,11 +92,25 @@ const ProfileDetails = (): JSX.Element => {
+          <Box display="flex" flexDirection="column" alignItems="center" justifyContent="center" paddingTop="10px">
+            {reviews.map((review) => {
+              return (
+                <CardContent key={review.clientId + review.sitterId} className={classes.review}>
```
Why don't we consider using something unique (id of a review document) because if an owner leaves more than one review - the respective component will have identical keys

rajivtitus commented on client/src/pages/ProfileDetails/ReviewForm/ReviewForm.tsx:
```diff
@@ -0,0 +1,52 @@
+      createReview(id, loggedInUser.id, rating, description).then((data) => {
+        if (data.success) {
+          updateSnackBarMessage('Review submitted successfully');
+          window.location.reload();
```
As opposed to reloading the entire window we can return the new review as a response and append it to our existing state

rajivtitus reviewed (commented):
Good work with this! Just made some comments for you to have a look and address :)

jb162 committed:
Small updates

addressed feedback in PR

jb162 committed:
Addressed Comments

added in function to add new review to reviews state var rather than reload page

jb162 committed:
Merge branch 'main' into justin-fs-reviews

jb162 committed:
last changes

last changes

rajivtitus commented on client/src/pages/ProfileDetails/BookingForm/BookingForm.tsx:
```diff
@@ -41,15 +41,25 @@ const BookingForm = (): JSX.Element => {
+        if (data.success) {
+          setAvailableTime(data.success.profile.availableTime);
+        } else {
+          console.log(data.error);
```
You can update the Snackbar with the error message rather than logging it to the console.

rajivtitus commented on client/src/pages/ProfileDetails/BookingForm/BookingForm.tsx:
```diff
@@ -41,15 +41,25 @@ const BookingForm = (): JSX.Element => {
+          avgRating = Math.round((avgRating * 2) / reviews.length) / 2;
+          setRating(avgRating);
+        } else {
+          console.log(data.error);
```
Ditto as comment above!

rajivtitus commented on client/src/pages/ProfileDetails/ReviewForm/ReviewForm.tsx:
```diff
@@ -1,26 +1,34 @@
 import { SnackBarContext } from '../../../context/useSnackbarContext';
 
-const ReviewForm = (): JSX.Element => {
+const ReviewForm = (props: any): JSX.Element => {
```
Please avoid using `any` as a type. Also, why don't we destructure the props? :)

jb162 replied:
I was having issues setting the type. I will go back and fix this tomorrow if I have time.

rajivtitus commented on client/src/pages/ProfileDetails/BookingForm/BookingForm.tsx:
```diff
@@ -35,10 +37,30 @@ const BookingForm = (): JSX.Element => {
   useEffect(() => {
     if (id) {
-      fetchProfile(id).then((data) => setAvailableTime(data.success.profile.availableTime));
+      fetchProfile(id).then((data) => {
```
Aren't we already fetching the sitter profile within the Profile Details page? We can just pass the `availableTime` as a prop to the booking form :)

jb162 replied:
I tried to change this but I was having issues with it and the previous code was fully functional so I just kept it so I could show it off at the presentation.

rajivtitus commented:
Thanks for making requested changes, just made a few comments - once addressed this PR should be good 👍🏽

jb162 committed:
last changes

last changes
